### PR TITLE
8335220: C2: Missing check for Opaque4 node in EscapeAnalysis

### PR DESCRIPTION
**Pull request by vnkozlov**
While testing Leyden Early Release the problem was found in C2 EscapeAnalysis, in new code which handles allocations merge. `ConnectionGraph:: can_reduce_check_users()` have incomplete checks which missed case with `Opaque4` node between `If` and `Bool` nodes. As result we hit assert in `ConnectionGraph::specialize_cmp()` when trying to access inputs of `Cmp` node
```
Node* curr_cmp = curr_ctrl->in(0)->in(1)->in(1); // true/false -> if -> bool -> cmp
con = curr_cmp->in(1)->is_Con() ? curr_cmp->in(1) : curr_cmp->in(2);
```

This code was introduce for ReduceAllocationMerges optimization [JDK-8316991](https://bugs.openjdk.org/browse/JDK-8316991) in JDK-23. There was followup fix [JDK-8330853](https://bugs.openjdk.org/browse/JDK-8330853) to add more checks but it missed this case.

I added more strict check in `can_reduce_check_users()` to catch such cases. And added asserts in `specialize_cmp()` to verify expected nodes.

Tested tier1-3.hs-stress,hs-xcomp

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335220](https://bugs.openjdk.org/browse/JDK-8335220): C2: Missing check for Opaque4 node in EscapeAnalysis (**Bug** - P3)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [a0687aee](https://git.openjdk.org/jdk/pull/19921/files/a0687aee60752a0c3cf08af8533072279c0e0e83)
 * [Cesar Soares Lucas](https://openjdk.org/census#cslucas) (@JohnTortugo - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19921/head:pull/19921` \
`$ git checkout pull/19921`

Update a local copy of the PR: \
`$ git checkout pull/19921` \
`$ git pull https://git.openjdk.org/jdk.git pull/19921/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19921`

View PR using the GUI difftool: \
`$ git pr show -t 19921`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19921.diff">https://git.openjdk.org/jdk/pull/19921.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19921#issuecomment-2193795852)

### CODE REVIEW

File: src/hotspot/share/opto/escape.cpp [hunk 1]
```diff
@@ -574,18 +574,20 @@ bool ConnectionGraph::can_reduce_check_users(Node* n, uint nesting) const {
         // CmpP/N used by the If controlling the cast.
         if (use->in(0)->is_IfTrue() || use->in(0)->is_IfFalse()) {
           Node* iff = use->in(0)->in(0);
-          if (iff->Opcode() == Op_If && iff->in(1)->is_Bool() && iff->in(1)->in(1)->is_Cmp()) {
+          bool can_reduce = (iff->Opcode() == Op_If) && iff->in(1)->is_Bool() && iff->in(1)->in(1)->is_Cmp();
+          if (can_reduce) {
             Node* iff_cmp = iff->in(1)->in(1);
             int opc = iff_cmp->Opcode();
-            if ((opc == Op_CmpP || opc == Op_CmpN) && !can_reduce_cmp(n, iff_cmp)) {
+            can_reduce = (opc == Op_CmpP || opc == Op_CmpN) && can_reduce_cmp(n, iff_cmp);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/escape.cpp:581]: Can you add a comment here that we could have an `Opaque4` node for which we want to bail out?

**Review by chhagedorn — APPROVED** [review]
Otherwise, looks good! Were you also able to reproduce this in mainline?

File: src/hotspot/share/opto/escape.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/escape.cpp:584]: Done

**Review by JohnTortugo — APPROVED** [review]
LGTM. Thank you for fixing.

**Comment by vnkozlov** [comment]
Thank you, Christian and Cesar, for reviews.

**Pull request closed without merging**